pandas/core/internals/array_manager.py

```diff
@@ -0,0 +1,902 @@
+    def _normalize_axis(axis):
+        # switch axis
+        axis = 1 if axis == 0 else 0
+        return axis
+
+    # TODO can be shared
+    def set_axis(self, axis: int, new_labels: Index) -> None:
+        # Caller is responsible for ensuring we have an Index object.
+        axis = self._normalize_axis(axis)
+        old_len = len(self._axes[axis])
+        new_len = len(new_labels)
+
+        if new_len != old_len:
+            raise ValueError(
+                f"Length mismatch: Expected axis has {old_len} elements, new "
+                f"values have {new_len} elements"
+            )
+
+        self._axes[axis] = new_labels
+
+    def consolidate(self) -> "ArrayManager":
+        return self
+
+    def is_consolidated(self) -> bool:
+        return True
+
+    def _consolidate_inplace(self) -> None:
+        pass
+
+    def get_dtypes(self):
+        return np.array([arr.dtype for arr in self.arrays], dtype="object")
+
+    # TODO setstate getstate
+
+    def __repr__(self) -> str:
+        output = type(self).__name__
+        output += f"\nIndex: {self._axes[0]}"
+        output += f"\nColumns: {self._axes[1]}"
+        output += f"\n{len(self.arrays)} arrays:"
+        for arr in self.arrays:
+            output += f"\n{arr.dtype}"
+        return output
+
+    def _verify_integrity(self) -> None:
+        n_rows, n_columns = self.shape_proper
+        if not len(self.arrays) == n_columns:
+            raise ValueError(
+                "Number of passed arrays must equal the size of the column Index: "
+                f"{len(self.arrays)} arrays vs {n_columns} columns."
+            )
+        for arr in self.arrays:
+            if not len(arr) == n_rows:
+                raise ValueError(
+                    "Passed arrays should have the same length as the rows Index: "
+                    f"{len(arr)} vs {n_rows} rows"
+                )
+            if not isinstance(arr, (np.ndarray, ExtensionArray)):
+                raise ValueError(
+                    "Passed arrays should be np.ndarray or ExtensionArray instances, "
+                    f"got {type(arr)} instead"
+                )
+
+    def reduce(
+        self: T, func: Callable, ignore_failures: bool = False
+    ) -> Tuple[T, np.ndarray]:
+        # TODO this still fails because `func` assumes to work on 2D arrays
+        # TODO implement ignore_failures
+        assert self.ndim == 2
+
+        res_arrays = []
+        for arr in self.arrays:
+            res = func(arr, axis=0)
+            res_arrays.append(np.array([res]))
+
+        index = Index([None])  # placeholder
+        new_mgr = type(self)(res_arrays, [index, self.items])
+        indexer = np.arange(self.shape[0])
+        return new_mgr, indexer
+
+    def operate_blockwise(self, other: "ArrayManager", array_op) -> "ArrayManager":
+        """
+        Apply array_op blockwise with another (aligned) BlockManager.
+        """
+        # TODO what if `other` is BlockManager ?
```

Comment:
prob need a .to_manager(kind=) method for each of these to convert to the other

Reply:
Yes, will add